### ISSUE TITLE
Add a vm.mockCallRevert cheatcode

### ISSUE THREAD

**Issue by fubhy** [foundry-rs/foundry#2740]
### Component

Forge

### Describe the feature you would like

I might be missing something but it seems like it's currently only possible to mock return values, but not reverts. For now, in my case I can simply create a Mock contract that does that for me but I'd imagine it could be handy to allow this via a cheatcode for convenience.

```solidity
vm.mockCallRevert("revert reason / error")
```

**Comment by onbjerg** [comment]
Makes sense

**Comment by mds1** [comment]
IMO if this is added we can also close https://github.com/foundry-rs/forge-std/issues/46, as cheatcodes would then cover the majority of mocking use cases

**Comment by tynes** [comment]
Could you mock a revert by returning `abi.encodeWithSignature("Error(string)", abi.encode("my error message"))` or if its a custom revert just abi encoding the custom error?

**Comment by mds1** [comment]
I don't think so, because the call's `success` output would still be true instead of false

**Comment by tynes** [comment]
I suppose you could make `vm.mockCall` aware of the data passed in and check to see if the 4byte selector matches `Error(string)` or any of the custom errors and have it return `false`, but that adds complexity to the implementation of mock call. I'd be happy as long as whatever solution could work with custom errors. Also wondering if something like this could be used to test exceptional errors like evm out of gas in a subcall

**Comment by onbjerg** [comment]
>Also wondering if something like this could be used to test exceptional errors like evm out of gas in a subcall

No, that would make the EVM unwind entirely. No way to stop that.

**Comment by tynes** [comment]
> > Also wondering if something like this could be used to test exceptional errors like evm out of gas in a subcall
> 
> No, that would make the EVM unwind entirely. No way to stop that.

I mean something like:

```solidity
vm.ExpectRevert(RevertReasons.OUT_OF_GAS);
(bool success, ) = target.call{ gas: 1 }(...);
```

Where you pass a defined amount of gas to a subcall and it reverts there but there is still gas left in the higher level call to continue. This is probably a more fringe usecase so not super important

**Comment by onbjerg** [comment]
This would still cause the EVM to unwind :)

**Comment by adhusson** [comment]
This would be very useful.

**Comment by sendra** [comment]
How is the state of this? It would be very useful

**Comment by reubenr0d** [comment]
I have a working mvp here: https://github.com/foundry-rs/foundry/pull/4343

Shall I take things forward?

**Comment by mds1** [comment]
That would be awesome, thanks @reubenr0d. I'll assign this to you. I think we need two functions, one that requires value, and one that doesn't:

```solidity
function mockCallRevert(address, bytes calldata, bytes calldata) external;
function mockCallRevert(address, uint256, bytes calldata, bytes calldata) external;
```

**Comment by reubenr0d** [comment]
Thanks @mds1, I've made an attempt here: https://github.com/foundry-rs/foundry/pull/4343
I'm still learning the ropes with rust and would appreciate a little help with this. The tests that `mockCallRevert` on an already `mockCall`ed call and vice versa fail right now.
